src/main/python/org/broadinstitute/hellbender/gatktool/tool.py

```diff
@@ -0,0 +1,175 @@
+
+
+def sendAck():
+    """
+    Send a positive acknowledgment to GATK. This should generally only
+    be called by python code that is embedded in Java, since the executor
+    keeps track of whether an ack reuest is outstanding.
+    """
+    global _ackFIFO
+    _ackFIFO.writeAck()
+
+
+def sendNack():
+    """
+    Send a negative acknowledgment to GATK. Generally only called by the
+    installed exception hook. This will result in a Java exception being
+    thrown that unless caught by Java code, will terminate the tool.
+    """
+    global _ackFIFO
+    _ackFIFO.writeNack()
+
+
+def terminateGATK():
+    """
+    Called by GATK when no more Python commands will be executed
+    """
+    global _ackFIFO
+    assert _ackFIFO != None
+    _ackFIFO.close()
+    _ackFIFO = None
+
+
+def initializeDataFIFO(dataFIFOName: str):
+    """
+    Initialize the data FIFO for reading.
+
+    Once this method has been called, the FIFO may be read using the
+    readDataFIFO function.
+    """
+    global _dataFIFO
+    _dataFIFO = DataFIFO(dataFIFOName)
+
+
+def closeDataFIFO():
+    """"
+    Close the data FIFO. Afteer this method hass been called, the
+    data FIFO can no longer be read.
+    """
+    global _dataFIFO
+    assert _dataFIFO != None
+    _dataFIFO.close()
+    _dataFIFO = None
+
+
+def readDataFIFO() -> str:
+    """
+    Read a line from the Data FIFO.
+    :return: string
+    """
+    global _dataFIFO
+    return _dataFIFO.readLine()
+
+
+class AckFIFO:
+    """
+    Manage the FIFO used to notify GATK (via an ack) that a command has
+    completed, or failed due to an unhandled exception (via a nck).
+    """
+    _ackString = "ack"
+    _nackString = "nck"
+
+    def __init__(self, ackFIFOName: str) -> None:
+        """Open the ack fifo stream for writing only"""
+        self.ackFIFOName = ackFIFOName
+        writeDescriptor = os.open(self.ackFIFOName, os.O_WRONLY)
+        self.fileWriter = os.fdopen(writeDescriptor, 'w')
+
+    def writeAck(self):
+        """
+        Write a positive acknowledgement to the ACK FIFO.
+        """
+        assert self.fileWriter != None
+        self.fileWriter.write(AckFIFO._ackString)
+        self.fileWriter.flush()
+
+    def writeNack(self):
+        """
+        Write a negative acknowledgement to the ACK FIFO.
+
+        Calling this method will result in an exception being thrown
+        in the GATK tool on whos behalf this module is running.
+        """
+        assert self.fileWriter != None
+        self.fileWriter.write(AckFIFO._nackString)
+        self.fileWriter.flush()
+
+    def close(self):
+        assert self.fileWriter != None
+        self.fileWriter.close()
+        self.fileWriter = None
+
+
+class DataFIFO:
+    """
+    Manage the FIFO stream used for transferring data from the GATK tool to
+    Python code.
+
+    The FIFO is written by GATK and read by Python.
+    """
+    # TODO: Need a way to provide a per-line deserializer
```

Comment:
Does this mean that each line is a new command?

Reply:
This was just a reminder for myself that ultimately we'll probably want a way to allow heterogenous data types to be passed through the FIFO (i.e., a variant followed by one or more reads), and if we used something structured like protobufs, we'll need some way to handle them on both sides (Java and Python). But for now this comment can be removed. Done.